posthog/temporal/batch_exports/redshift_batch_export.py

```diff
@@ -78,22 +79,104 @@ def remove_escaped_whitespace_recursive(value):
+
+        Keeps track of the current cursor_factory to set it after we are done.
+        """
+        current_factory = self.connection.cursor_factory
+        self.connection.cursor_factory = psycopg.AsyncClientCursor
+
+        try:
+            async with self.connection.cursor() as cursor:
+                # Not a fan of typing.cast, but we know this is an psycopg.AsyncClientCursor
+                # as we have just set cursor_factory.
+                cursor = typing.cast(psycopg.AsyncClientCursor, cursor)
+                yield cursor
+        finally:
+            self.connection.cursor_factory = current_factory
+
+    async def amerge_identical_tables(
+        self,
+        final_table_name: str,
+        stage_table_name: str,
+        schema: str,
+        merge_key: Fields,
+        update_when_matched: Fields,
+        version_key: str = "version",
+    ) -> None:
+        """Merge two identical tables in PostgreSQL."""
+        if schema:
+            final_table_identifier = sql.Identifier(schema, final_table_name)
+            stage_table_identifier = sql.Identifier(schema, stage_table_name)
 
-    This just yields a Postgres connection but we adjust a couple of things required for
-    psycopg to work with Redshift:
-    1. Set UNICODE encoding to utf-8 as Redshift reports back UNICODE.
-    2. Set prepare_threshold to None on the connection as psycopg attempts to run DEALLOCATE ALL otherwise
-        which is not supported on Redshift.
-    """
-    psycopg._encodings._py_codecs["UNICODE"] = "utf-8"
-    psycopg._encodings.py_codecs.update((k.encode(), v) for k, v in psycopg._encodings._py_codecs.items())
+        else:
+            final_table_identifier = sql.Identifier(final_table_name)
+            stage_table_identifier = sql.Identifier(stage_table_name)
+
+        and_separator = sql.SQL("AND")
+        merge_condition = and_separator.join(
+            sql.SQL("{final_field} = {stage_field}").format(
+                final_field=sql.Identifier(schema, final_table_name, field[0]),
+                stage_field=sql.Identifier("stage", field[0]),
+            )
+            for field in merge_key
+        )
+
+        delete_condition = and_separator.join(
+            sql.SQL("{final_field} = {stage_field}").format(
+                final_field=sql.Identifier("final", field[0]),
+                stage_field=sql.Identifier(schema, stage_table_name, field[0]),
+            )
+            for field in merge_key
+        )
+
+        delete_query = sql.SQL("""\
+        DELETE FROM {stage_table}
+        USING {final_table} AS final
+        WHERE {merge_condition} AND {stage_table}.{stage_version_key} > final.{final_version_key};
+        """).format(
+            final_table=final_table_identifier,
+            stage_table=stage_table_identifier,
+            merge_condition=delete_condition,
+            stage_version_key=sql.Identifier(version_key),
+            final_version_key=sql.Identifier(version_key),
+        )
```

Comment:
Redshift doesn't support adding a condition on the merge, so we have to delete any existing rows here

Reply:
Also, redshift doesn't support aliasing tables sometimes, it's very annoying.